### PR TITLE
fix: rewrite benchmark reporter without `log-update`

### PR DESCRIPTION
**Pull request by AriPerkkio**
### Description

- Fixes https://github.com/vitest-dev/vitest/issues/5895
- ~~Patches `tinyexec` for https://github.com/tinylibs/tinyexec/issues/40.~~ Alternatively I could remove the patch and use `node:child_process` in test cases.
- Removes `log-update` and `cli-truncate` -> -92 kB from bundle
- Uses summary-reporter so users can now scroll up and see previous benchmark results while tests are running.

### Test preview release with:
- [x] https://github.com/radashi-org/radashi/tree/main/benchmarks
  - They have 98 benchmark files. In `vitest@v2` they could not see previous results during the run as `log-update` limits the rows to the terminal height. Now you can actually scroll up during the run and see previous results. No need to wait for that ~8min run to finish.
  - Having the summary reporter on bottom of terminal helps here a lot as you can actually see how many benchmark files are left.
- [x] https://github.com/toss/es-toolkit/tree/main/benchmarks/performance
  - Noticed tests being reported as `'pass'` multiple times, fixed. 
- [x] https://github.com/vuejs/core/blob/main/packages/reactivity/__benchmarks__/ref.bench.ts

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/vitest/src/node/reporters/renderers/utils.ts [hunk 1]
```diff
@@ -258,3 +236,11 @@ export function withLabel(color: 'red' | 'green' | 'blue' | 'cyan' | 'yellow', l
...
   return c.dim(`${str.padStart(11)} `)
 }
+
+export function truncateString(text: string, maxLength: number): string {
+  if (stripVTControlCharacters(text).length <= maxLength) {
+    return text
+  }
+
+  return `${sliceAnsi(text, 0, maxLength - 1)}…`
+}
```
Comment (AriPerkkio) [inline packages/vitest/src/node/reporters/renderers/utils.ts:246]: `slice-ansi` is only required here. We could remove it if we didn't care about coloring when error messages' lines are truncated. Thoughts @hi-ogawa, @sheremet-va?

image

Comment (hi-ogawa) [reply packages/vitest/src/node/reporters/renderers/utils.ts:246]: I don't mind either way, but unless it's buggy or slow, I think It's good to have it.

Comment (AriPerkkio) [reply packages/vitest/src/node/reporters/renderers/utils.ts:246]: Not buggy or slow. I was mostly thinking about reducing install size, but maybe it's not worth it. Though this truncating is only done for these errors so it's not used most of the time. https://pkg-size.dev/slice-ansi

Comment (sheremet-va) [reply packages/vitest/src/node/reporters/renderers/utils.ts:246]: I also don't mind

File: packages/vitest/src/node/logger.ts [hunk 2]
```diff
@@ -303,6 +307,42 @@ export class Logger {
     this.log(c.red(divider()))
   }
 
+  getColumns() {
+    return 'columns' in this.outputStream ? this.outputStream.columns : 80
+  }
+
+  onExit(listener: Listener) {
```
Comment (sheremet-va) [inline packages/vitest/src/node/logger.ts:314]: From the outside perspective, I am a bit confused by the `logger` is responsible for `onExit` callback 👀

Comment (sheremet-va) [reply packages/vitest/src/node/logger.ts:314]: I like `onTerminalCleanup` 👍🏻

File: packages/vitest/src/node/reporters/renderers/utils.ts (hunk 1 above)
Comment (AriPerkkio) [reply packages/vitest/src/node/reporters/renderers/utils.ts:246]: I'll remove `slice-ansi` from here as its usage is quite an edge case.

**Comment by AriPerkkio** [comment]
Looks like benchmark runner is reporting tests of nested suites twice. Need to resolve this, otherwise summary is showing too high test count.

**Comment by sheremet-va** [comment]
Does this also add a project name to the output? I just noticed we don't print `|name|`

**Comment by AriPerkkio** [comment]
Project name is visible now:

**Comment by AriPerkkio** [comment]
It's not in table output though, should it be here too?

**Comment by sheremet-va** [comment]
> It's not in table output though, should it be here too?
> 
> 

Yeah

File: packages/vitest/src/runtime/runners/benchmark.ts [hunk 3]
```diff
@@ -135,16 +137,6 @@ async function runBenchmarkSuite(suite: Suite, runner: NodeBenchmarkRunner) {
     suite.result!.duration = performance.now() - start
     suite.result!.state = 'pass'
 
-    tasks
```
Comment (sheremet-va) [inline packages/vitest/src/runtime/runners/benchmark.ts:138]: Did you move the sorting? Why is it removed from here?

Comment (AriPerkkio) [reply packages/vitest/src/runtime/runners/benchmark.ts:138]: It's moved to main thread's side, so that we don't need to report tasks twice as `'pass'`.

- Created failing test case first: https://github.com/vitest-dev/vitest/pull/7019/commits/d4223fd55f46c9ef34d0e82326da8e523a110d3b
- Then moved sorting to the other side of RPC: https://github.com/vitest-dev/vitest/pull/7019/commits/9ed93b56752033409c4146c5e1a69c0d94281830

There's `test/benchmark/test/reporter.test.ts` that fails if this sorting and `result.benchmark.rank` setting is wrong.

**Review by hi-ogawa — APPROVED** [review]
Looks good! Tested radashi locally and the improvement is awesome :clap: 

About `tinyexec` patch, it's not a blocker, but the situation is not ideal since patch is only applied for our local dev, but `tinyexec` is also used for external dependencies in a few places.

**Comment by AriPerkkio** [comment]
> About `tinyexec` patch, it's not a blocker, but the situation is not ideal since patch is only applied for our local dev, but `tinyexec` is also used for external dependencies in a few places.

I agree. I'll replace `tinyexec` usage+patching with `node:child_process` or maybe even `startVitest` when using `preserveAnsi` option from https://github.com/vitest-dev/vitest/pull/7090/files#diff-eee947831abd0685dc52e26f42f2c7c3adfb802e478ca8e252ba649f1a1ab2f6R5-R13.

**Review by hi-ogawa — APPROVED** [review]
Thanks!
